src/librustc_mir/transform/simplify_comparison_integral.rs

```diff
@@ -0,0 +1,252 @@
+        },
+        _ => None,
+    }
+}
+
+struct OptimizationFinder<'a, 'tcx> {
+    body: &'a Body<'tcx>,
+}
+
+impl<'a, 'tcx> OptimizationFinder<'a, 'tcx> {
+    fn find_optimizations(&self) -> Vec<OptimizationInfo<'tcx>> {
+        self.body
+            .basic_blocks()
+            .iter_enumerated()
+            .filter_map(|(bb_idx, bb)| {
+                // find switch
+                let (place_switched_on, values, targets) = match &bb.terminator().kind {
+                    rustc_middle::mir::TerminatorKind::SwitchInt {
+                        discr, values, targets, ..
+                    } => Some((discr.place()?, values, targets)),
+                    _ => None,
+                }?;
+
+                // find the statement that assigns the place being switched on
+                bb.statements.iter().enumerate().rev().find_map(|(stmt_idx, stmt)| {
+                    match &stmt.kind {
+                        rustc_middle::mir::StatementKind::Assign(box (lhs, rhs))
+                            if *lhs == place_switched_on =>
+                        {
+                            match rhs {
+                                Rvalue::BinaryOp(op, left, right)
+                                    if *op == BinOp::Eq || *op == BinOp::Ne =>
+                                {
+                                    // check that either left or right is a constant.
+                                    // if any are, we can use the other to switch on, and the constant as a value in a switch
+                                    use Operand::*;
+                                    match (left, right) {
+                                        (
+                                            Constant(branch_value),
+                                            Copy(to_switch_on) | Move(to_switch_on),
+                                        )
+                                        | (
+                                            Copy(to_switch_on) | Move(to_switch_on),
+                                            Constant(branch_value),
+                                        ) => {
+                                            let branch_value_ty = branch_value.literal.ty;
+
+                                            let branch_value_scalar =
+                                                branch_value.literal.val.try_to_scalar().or_else(
+                                                    || {
+                                                        trace!(
+                                                            "NO: constant value {:?} not a scalar",
+                                                            branch_value
+                                                        );
+                                                        None
+                                                    },
+                                                )?;
```

Comment:
If you don't need the `trace!` anymore, you can also write 

```suggestion
                                                branch_value.literal.val.try_to_scalar().ok()?;
```